core/shared/src/main/scala/zio/Schedule2.scala

```diff
@@ -0,0 +1,1009 @@
+      else Some(Interval(istart, iend))
+    }
+  }
+  object Interval extends Function2[Instant, Instant, Interval] {
+
+    /**
+     * Constructs a new interval from the two specified endpoints. If the start endpoint greater
+     * than the end endpoint, then a zero size interval will be returned.
+     */
+    def apply(start: Instant, end: Instant): Interval =
+      if (start.isBefore(end) || start == end) new Interval(start, end) {}
+      else empty
+
+    def after(instant: Instant): Interval = Interval(instant, Instant.MAX)
+
+    def before(instant: Instant): Interval = Interval(Instant.MIN, instant)
+
+    /**
+     * An interval of zero-width.
+     */
+    val empty: Interval = Interval(Instant.ofEpochSecond(0L), Instant.ofEpochMilli(0L))
+
+    def fromInstantDuration(instant: Instant, duration: Duration): Interval =
+      Interval(instant, instant.plusNanos(duration.toNanos))
+
+    private def min(l: Instant, r: Instant): Instant = if (l.compareTo(r) <= 0) l else r
+    private def max(l: Instant, r: Instant): Instant = if (l.compareTo(r) >= 0) l else r
+  }
+
+  type StepFunction[-Env, -In, +Out] = (Instant, In) => ZIO[Env, Nothing, Schedule2.Decision[Env, In, Out]]
+  object StepFunction {
+    def done[A](a: => A): StepFunction[Any, Any, A] = (_: Instant, _: Any) => ZIO.succeed(Decision.Done(a))
+
+    def stepIfNecessary[Env, In, Out](
+      now: Instant,
+      in: In,
+      option: Option[(Interval, Out)],
+      step0: StepFunction[Env, In, Out]
+    ): ZIO[Env, Nothing, Decision[Env, In, Out]] =
+      option match {
+        case None => step0(now, in)
+        case Some((interval, out)) =>
+          if (now.compareTo(interval.end) >= 0) step0(now, in)
+          else ZIO.succeed(Decision.Continue(out, interval, step0))
+      }
+  }
+
+  sealed trait Decision[-Env, -In, +Out] { self =>
+    def out: Out
+
+    final def as[Out2](out2: => Out2): Decision[Env, In, Out2] = map(_ => out2)
+
+    final def contramap[In1](f: In1 => In): Decision[Env, In1, Out] =
+      self match {
+        case Decision.Done(v) => Decision.Done(v)
+        case Decision.Continue(v, i, n) =>
+          Decision.Continue(v, i, (now: Instant, in1: In1) => n(now, f(in1)).map(_.contramap(f)))
+      }
+
+    final def map[Out2](f: Out => Out2): Decision[Env, In, Out2] =
+      self match {
+        case Decision.Done(v) => Decision.Done(f(v))
+        case Decision.Continue(v, i, n) =>
+          Decision.Continue(f(v), i, (now: Instant, in: In) => n(now, in).map(_.map(f)))
+      }
+
+    final def toDone: Decision[Env, Any, Out] =
+      self match {
+        case Decision.Done(v)           => Decision.Done(v)
+        case Decision.Continue(v, _, _) => Decision.Done(v)
+      }
+  }
+  object Decision {
+    final case class Done[-Env, +Out](out: Out) extends Decision[Env, Any, Out]
+    final case class Continue[-Env, -In, +Out](
+      out: Out,
+      interval: Interval,
+      next: StepFunction[Env, In, Out]
```

Comment:
or perhaps
```scala
type StepFunction[-Env, -I, +O] = UIO[I => ZIO[Env, Nothing, Decision[Env, I, O]]]
```

Reply:
The problem is if you call `unfold` twice, with the same params, you should get back the same schedule. But in this case, you will get back two different schedules, because the creation of the mutable variable is done as part of the (supposedly pure) function invocation. The two purely functional encodings are:

1. Explicit state, like the current `Schedule`, or maybe stored in a `Ref`.
2. Implicit state, which involves using "successor" step functions like `next`. 

After working with both, I think the second one is cleaner. I think I can make combinators to minimize duplication, but never having to worry about explicit state is quite nice.

Reply:
Fair enough 👌